tfx/examples/imdb/imdb_utils_native_keras.py

```diff
@@ -0,0 +1,210 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python source file include IMDB pipeline functions and necessary utils.
+
+The utilities in this file are used to build a model with native Keras.
+This module file will be used in Transform and generic Trainer."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from typing import List, Text
+
+import tensorflow as tf
+from tensorflow import keras
+import tensorflow_transform as tft
+
+from tfx.components.trainer.executor import TrainerFnArgs
+
+# There are 100 entries in the imdb_small dataset. ExampleGen splits the dataset
+# with a 2:1 train-eval ratio. Batch_size is an empirically sound
+# configuration.
+# To train on the entire imdb dataset, please refer to imdb_dataset_utils.py
+# and change the batch configuration accordingly.
+_DROPOUT_RATE = 0.2
+_EVAL_BATCH_SIZE = 5
+_HIDDEN_UNITS = 21
+_LABEL_KEY = "label"
+_LEARNING_RATE = 1e-4
+_MAX_FEATURES = 8000
+_MAX_LEN = 128
+_TRAIN_BATCH_SIZE = 10
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _tokenize_review(review):
+  """Tokenize the reviews by spliting the reviews, then constructing a
+  vocabulary. Map the words to their frequency index in the vocabulary."""
+  review_sparse = tf.strings.split(tf.reshape(review, [-1])).to_sparse()
+  # tft.apply_vocabulary doesn't reserve 0 for oov words. In order to comply
+  # with convention and use mask_zero in keras.embedding layer, manually set
+  # default value to -1 and add 1 to every index.
+  review_indices = tft.compute_and_apply_vocabulary(
+      review_sparse,
+      default_value=-1,
+      top_k=_MAX_FEATURES)
+  dense = tf.sparse.reset_shape(review_indices, None)
+  dense = tf.sparse.to_dense(review_indices, default_value=-1)
+  # TFX transform expects the transform result to be FixedLenFeature.
+  padding_config = [[0, 0], [0, _MAX_LEN]]
+  dense = tf.pad(
+      dense,
+      padding_config,
+      'CONSTANT',
+      -1)
+
+  padded = tf.slice(dense, [0, 0], [-1, _MAX_LEN])
+  padded += 1
+  return padded
+
+# TFX Transform will call this function.
+def preprocessing_fn(inputs):
+  """tf.transform's callback function for preprocessing inputs.
+
+  Args:
+    inputs: map from feature keys to raw not-yet-transformed features.
+
+  Returns:
+    Map from string feature key to transformed feature operations.
+  """
+  label = inputs['label']
```

Comment:
nit: we can remove this and use inputs['label'] and inputs['text'] directly.

Reply:
fixed